### PR TITLE
Add aosm extension

### PR DESCRIPTION
**Pull request by sunnycarter**
---
# Summary
Work item https://msazure.visualstudio.com/One/_workitems/edit/17617678 on AZ CLI team for the CLI Engagement Onboard.
`az aosm` commands for Azure Operator Service Manager.

GA Launch Date: 2023-09-30

This is our first Az CLI release and we intend to release the CLI with version <1.0.0. 

This is the first time that our team has gone through the APEX process so please can you lay out the next steps after this review so that we can make sure we understand correctly?

Our swagger definition:
-  this tool currently targets and is built against [https://github.com/Azure/azure-rest-api-specs-pr/tree/RPSaaSMaster/specification/hybridnetwork/resource-manager/Microsoft.HybridNetwork/preview/2023-04-01-preview](https://github.com/Azure/azure-rest-api-specs-pr/tree/RPSaaSMaster/specification/hybridnetwork/resource-manager/Microsoft.HybridNetwork/preview/2023-04-01-preview)
- for release, the tool will target 2023-09-01 API version, in **public** github repo.  The API is not currently there, the private version is currently at [https://github.com/Azure/azure-rest-api-specs-pr/tree/RPSaaSMaster/specification/hybridnetwork/resource-manager/Microsoft.HybridNetwork/stable/2023-09-01](https://github.com/Azure/azure-rest-api-specs-pr/tree/RPSaaSMaster/specification/hybridnetwork/resource-manager/Microsoft.HybridNetwork/stable/2023-09-01). **There is still work for us to do here before final review and merge.**

We have had an initial meeting with Damien Caro where we explained that standard CRUD operations on many of the HybridNetwork resources are not very beneficial to the user, as it is more likely that they will want to do a suite of complex operations involving multiple linked resource types.  As such, have written a customization for the Azure CLI, as an extension `az aosm` which will help them do this. This does NOT include every CRUD operation you can do over the API, Peter Whiting from AOSM team has been in discussions with the Jeremy Li in the CLI team about this.

I am asking AZ CLI team to make a second review of this extension. **We want to submit the PR finally 15th September 2023.**

Kind Regards and thanks in advance,

Sunny Carter 

## More detail
The readme outlines the commands, let me expand a little more on the design.

The basic overview of the function is explained here: [azure-cli-extensions/README.md](https://github.com/Azure/azure-cli-extensions/pull/6426/files#diff-0cfcd8eec109417e257251b623a5ecf0955ac0bfccd03eeaa7df3ab2861825ff) . To add more detail to that, our customization would:

-	Output one or more bicep templates which would create the most complex resources in our API
-	Allow the user to review those templates
-	When the user chooses to publish, create some of the simpler HybridNetwork resources using the generated Python SDK. These being:

    * Publisher
    * Network Function Definition Group
    * Artifact Stores (these can be backed by an ACR or by a storage account)
    * Network Service Design Group
    * Site
    
- Then deploy the bicep templates for the more complex HybridNetwork resources
   * Upload binary artifacts to the artifact stores using either  Blob storage Python API for Azure storage account, or ORAS python implementation for ACR

This checklist is used to make sure that common guidelines for a pull request are followed.

### General Guidelines

- [todo ] Have you run `azdev style <YOUR_EXT>` locally? (`pip install azdev` required)
Yes
- [ todo] Have you run `python scripts/ci/test_index.py -q` locally?
Yes. output is 
```
Ran 9 tests in 0.019s

OK (skipped=2)
```
Tests:
See [src/aosm/development.md](https://github.com/Azure/azure-cli-extensions/blob/df1ca06075f90a7387e54fc66611e7af985aa172/src/aosm/development.md)
We have Unit Tests and Integration tests.  The integration tests seem to be failing with an intermittent issue where the the test is looking for a request that exist in the recording of a test multiple times while it was only run once in the live test

This issue is intermittent but can be fixed by running a test live or allowing playback repeats in the cassette. We did not have enough time to test those solutions in our branch. 

I have spoken to Zelin Wang from the Azure CLI team about this and he confirmed that this was an issue that they have seen before too. Here are his recommendations:

_

> Found a similar issues in our repo:
> https://github.com/Azure/azure-cli/issues/19804 
> 
> And the fix pr:
> https://github.com/Azure/azure-cli/pull/19848/files (change to live only)
> Another similar PR that fix the same issue: https://github.com/Azure/azure-cli/pull/27016
> Root cause: https://github.com/Azure/azure-cli/pull/27016#issuecomment-1652933033
> 
> -------------------------------------------------------------------------------------------------
> 
> both MainThread and LROPoller are consuming the same URL https://cli-test-kv-ct-sd-000002.vault.azure.net/certificates/cert1/pending?api-version=REDACTED and waiting for a response with "status":"completed"
> 
> If MainThread runs faster, get_certificate_operation in LROPoller thread will not have enough recording entries to consume and fail, but the error won't surface because the result of the poller is not checked.
> If LROPoller thread runs faster, get_certificate_operation in MainThread thread will not have enough recording entries to consume and fail. An error will be thrown.

_

For new extensions:

- [yes ] My extension description/summary conforms to the [Extension Summary Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/extensions/extension_summary_guidelines.md).

### About Extension Publish

There is a pipeline to automatically build, upload and publish extension wheels.  
Once your pull request is merged into main branch, a new pull request will be created to update `src/index.json` automatically.  
You only need to update the version information in file setup.py and historical information in file HISTORY.rst in your PR but do not modify `src/index.json`.

### CODE REVIEW

**Comment by kairu-ms** [comment]
Hi @zhoxing-ms, could you help review this PR as it using bicep template.

File: .github/workflows/BuildAOSMWheel.yml
Comment (jsntcy) [inline .github/workflows/BuildAOSMWheel.yml:14]: Could you please tell why we need this github action to build wheel? Our CI will build wheel file for your extension automatically

File: .github/workflows/CheckStyleAndLinting.yml
Comment (jsntcy) [inline .github/workflows/CheckStyleAndLinting.yml:12]: The same question as above. Our CI will check style and linter for your extension automatically.

File: src/aosm/HISTORY.rst
Comment (zhoxing-ms) [inline src/aosm/HISTORY.rst:6]: Suggest keeping consistency with the style from other modules (such as `aks`)

```suggestion
upcoming
```

File: src/aosm/setup.py
Comment (zhoxing-ms) [inline src/aosm/setup.py:21]: Just out of curiosity, why did you directly upgrade to `0.2.0` version but skip `0.1.0` version in this initial extension version?

File: src/aosm/setup.py
Comment (zhoxing-ms) [inline src/aosm/setup.py:31]:
```suggestion
```

File: src/aosm/azext_aosm/commands.py
Comment (zhoxing-ms) [inline src/aosm/azext_aosm/commands.py:17]: Do you need to consider adding `confirmation=true` for the delete command?

File: src/aosm/azext_aosm/deploy/deploy_with_arm.py
Comment (zhoxing-ms) [inline src/aosm/azext_aosm/deploy/deploy_with_arm.py:500]: Is there really a situation where this extension is installed but Azure CLI is not installed?

File: src/aosm/azext_aosm/deploy/deploy_with_arm.py
Comment (zhoxing-ms) [inline src/aosm/azext_aosm/deploy/deploy_with_arm.py:459]: Please use `LongRunningOperation` instead of `.result()`, which can support the progress bar by default for the long polling operations
```suggestion
        # Wait for the deployment to complete and get the outputs
        deployment: DeploymentExtended = LongRunningOperation(cmd.cli_ctx)(validation_poller)
```

File: src/aosm/azext_aosm/deploy/deploy_with_arm.py
Comment (zhoxing-ms) [inline src/aosm/azext_aosm/deploy/deploy_with_arm.py:424]:
```suggestion
        validation_res = LongRunningOperation(cmd.cli_ctx)(validation)
```

File: src/aosm/azext_aosm/deploy/deploy_with_arm.py
Comment (zhoxing-ms) [reply src/aosm/azext_aosm/deploy/deploy_with_arm.py:459]: I see many similar issues elsewhere, so I won't repeat my comments

File: src/aosm/setup.py
Comment (sunnycarter) [reply src/aosm/setup.py:21]: Because we have sent some internal people the wheel to have a play with the tool before it gets merged.  We originally sent them 0.1.0 but made a breaking change, so updated to 0.2.0

File: .github/workflows/CheckStyleAndLinting.yml
Comment (sunnycarter) [reply .github/workflows/CheckStyleAndLinting.yml:12]: I have removed this file now (left it on our pre-release branch only)

File: src/aosm/azext_aosm/deploy/deploy_with_arm.py
Comment (sunnycarter) [reply src/aosm/azext_aosm/deploy/deploy_with_arm.py:640]: Removed this section of code

File: src/aosm/azext_aosm/deploy/deploy_with_arm.py
Comment (sunnycarter) [reply src/aosm/azext_aosm/deploy/deploy_with_arm.py:598]: I have changed them all

File: src/aosm/azext_aosm/deploy/deploy_with_arm.py
Comment (sunnycarter) [reply src/aosm/azext_aosm/deploy/deploy_with_arm.py:559]: I have changed all examples of this

File: src/aosm/HISTORY.rst
Comment (yanzhudd) [inline src/aosm/HISTORY.rst:6]: if it should be a new version instead of `upcoming`?

File: src/aosm/setup.py
Comment (jsntcy) [reply src/aosm/setup.py:19]: Only when the extension is officially released, then we need upgrade the version; otherwise we should not upgrade the version in case it confuses customers.

File: src/aosm/HISTORY.rst
Comment (jsntcy) [inline src/aosm/HISTORY.rst:38]: Make release history align with the version as I commented below.

File: src/aosm/setup.py
Comment (jsntcy) [reply src/aosm/setup.py:19]: As this is the first release of the extension, the version should be 1.0,0b1 per our versioning strategy.

**Comment by jsntcy** [comment]
Please fix CI failure first.

**Comment by sunnycarter** [comment]
> 

@jsntcy as explained in the overview and my email, we have some problems with the integration tests and would like a review before they are fixed.  Did your comment mean please fix CI failures before you review?  Or please fix CI failures before we release?

We intend to fix them and submit for final review before release.

I would strongly request a review before fixing the integration tests. I also do not know what to do to fix the license CI failure, could you give some pointers?

**Comment by PerfectChaos** [comment]
@yanzhudd @jsntcy - All CI issues have been resolved and checks are now passing, and per comments above I've cleaned out the version history to reflect the one version we're releasing: 1.0.0b1.

Our target GA date draws ever closer - would it be possible to get the final review of this ASAP?

File: src/aosm/azext_aosm/custom.py
Comment (yanzhudd) [inline src/aosm/azext_aosm/custom.py:107]: this function does not have a parameter `definition_type`
```suggestion
```

File: src/aosm/azext_aosm/custom.py
Comment (yanzhudd) [inline src/aosm/azext_aosm/custom.py:99]: the `config` variable might be referenced before assignment, since `definition_type` parameter might be None.

File: src/aosm/azext_aosm/custom.py
Comment (yanzhudd) [inline src/aosm/azext_aosm/custom.py:68]: just a reminder, the parameter of `order_params` is missing from the function documentation.

File: src/aosm/azext_aosm/_params.py
Comment (yanzhudd) [inline src/aosm/azext_aosm/_params.py:29]: is this param required for users when using all `aosm nfd` commands? If so, please add `required=True` for this param.
After adding this restriction, combining with `arg_type=definition_type`, users can only input specific values, and it is not necessary to validate the users' input like https://github.com/jddarby/azure-cli-extensions/blob/c172f412bf91516a726f1ec74ff8ee3ca5a9b534/src/aosm/azext_aosm/custom.py#L252.

**Comment by yanzhudd** [comment]
may I ask if you have tested `aosm nfd/nsd generate-config` command?

File: src/aosm/azext_aosm/custom.py
Comment (jamiedparsons) [reply src/aosm/azext_aosm/custom.py:368]: No the print goes to console and the log goes to the log file, so it isn't duplicated.

**Comment by jamiedparsons** [comment]
> may I ask if you have tested `aosm nfd/nsd generate-config` command?

Yes, we have manually tested and have UTs:
https://github.com/jddarby/azure-cli-extensions/blob/add-aosm-extension/src/aosm/azext_aosm/tests/latest/test_cnf.py#L17
https://github.com/jddarby/azure-cli-extensions/blob/add-aosm-extension/src/aosm/azext_aosm/tests/latest/test_nsd.py#L269
https://github.com/jddarby/azure-cli-extensions/blob/add-aosm-extension/src/aosm/azext_aosm/tests/latest/test_vnf.py#L17

File: .flake8
Comment (zhoxing-ms) [inline .flake8:17]: Why do you need to modify this file?

File: src/aosm/azext_aosm/delete/delete.py
Comment (zhoxing-ms) [inline src/aosm/azext_aosm/delete/delete.py:65]:
```suggestion
                print("There is no undo. Type the publisher name to confirm.")
```

File: src/aosm/azext_aosm/delete/delete.py
Comment (zhoxing-ms) [inline src/aosm/azext_aosm/delete/delete.py:140]: In fact, the design of the vast majority of CLI delete commands defaults to successful deletion if no information is output. Therefore, it is recommended to keep the design style consistent with other CLI commands.

File: src/aosm/azext_aosm/delete/delete.py
Comment (zhoxing-ms) [inline src/aosm/azext_aosm/delete/delete.py:155]: Why print and record debug logs simultaneously?

File: src/aosm/azext_aosm/delete/delete.py
Comment (zhoxing-ms) [inline src/aosm/azext_aosm/delete/delete.py:165]: Same as https://github.com/Azure/azure-cli-extensions/pull/6426/files#r1366418669

File: src/aosm/azext_aosm/delete/delete.py
Comment (zhoxing-ms) [inline src/aosm/azext_aosm/delete/delete.py:200]: Same as https://github.com/Azure/azure-cli-extensions/pull/6426/files#r1366420334

File: src/aosm/azext_aosm/delete/delete.py
Comment (zhoxing-ms) [inline src/aosm/azext_aosm/delete/delete.py:225]: Same as above. There are many similar places below, so I won't repeat the comments.

File: src/aosm/azext_aosm/delete/delete.py
Comment (zhoxing-ms) [inline src/aosm/azext_aosm/delete/delete.py:235]: Same as above. There are many similar places below, so I won't repeat the comments.

File: src/aosm/setup.py
Comment (zhoxing-ms) [inline src/aosm/setup.py:34]: Just out of curiosity, can the dependency `azure-storage-blob` be included by the dependency `azure-multiapi-storage` and `azure-storage-common` in the azure-cli-core? [code link](https://github.com/Azure/azure-cli/blob/be065f1cab2be4549c9d5d4e57cd0e69c904f441/src/azure-cli/setup.py#L128-L129)

**Comment by zhoxing-ms** [comment]
Due to the super huge size of this PR, and CLI team only reviews the code specification and is not responsible for reviewing the business logic, please have at least one engineer from your team review and sign off this PR

File: src/aosm/azext_aosm/delete/delete.py
Comment (Cyclam) [reply src/aosm/azext_aosm/delete/delete.py:65]: Thanks, done.

File: src/aosm/azext_aosm/custom.py
Comment (Cyclam) [reply src/aosm/azext_aosm/custom.py:103]: Fixed

File: src/aosm/azext_aosm/custom.py
Comment (Cyclam) [reply src/aosm/azext_aosm/custom.py]: Fixed

File: src/aosm/azext_aosm/_params.py
Comment (Cyclam) [reply src/aosm/azext_aosm/_params.py]: Fixed

File: .flake8
Comment (Cyclam) [reply .flake8]: The linter complained about the existing format (inline comments and missing commas), so it was modified to keep the linter happy. But you're right, this isn't our file to worry about, so I've reverted it to how it was.

File: src/aosm/azext_aosm/delete/delete.py
Comment (Cyclam) [reply src/aosm/azext_aosm/delete/delete.py]: I've removed the message. Thanks for the advice.

File: src/aosm/azext_aosm/delete/delete.py
Comment (Cyclam) [reply src/aosm/azext_aosm/delete/delete.py:155]: We need to rethink how we output to console. These changes are too widespread to make before GA release, so I've opened work item https://dev.azure.com/msazuredev/AzureForOperators/_workitems/edit/992775 to track.

File: src/aosm/azext_aosm/delete/delete.py
Comment (Cyclam) [reply src/aosm/azext_aosm/delete/delete.py:200]: We need to rethink how we output to console. These changes are too widespread to make before GA release, so I've opened work item https://dev.azure.com/msazuredev/AzureForOperators/_workitems/edit/992775 to track.

File: src/aosm/azext_aosm/delete/delete.py
Comment (Cyclam) [reply src/aosm/azext_aosm/delete/delete.py:225]: We need to rethink how we output to console. These changes are too widespread to make before GA release, so I've opened work item https://dev.azure.com/msazuredev/AzureForOperators/_workitems/edit/992775 to track.

File: src/aosm/azext_aosm/delete/delete.py
Comment (Cyclam) [reply src/aosm/azext_aosm/delete/delete.py]: I've moved the message to logger.info(), so won't be output by default. Thanks for the advice.

File: src/aosm/azext_aosm/delete/delete.py
Comment (Cyclam) [reply src/aosm/azext_aosm/delete/delete.py]: I've moved all similar messages to logger.info(), so won't be output by default. Thanks for the advice.

**Comment by Cyclam** [comment]
> Due to the super huge size of this PR, and CLI team only reviews the code specification and is not responsible for reviewing the business logic, please have at least one engineer from your team review and sign off this PR

Thanks for your review.

The business logic has been developed over the last few months, with every merge into our development branch reviewed by at least one member of the team.

Several of us have also done a general high level review before this PR, and we can sign off the business logic as fully reviewed.

File: src/aosm/setup.py
Comment (Cyclam) [reply src/aosm/setup.py:34]: `azure-storage-blob` depends on `azure-storage-common`  (not vice versa).

https://pypi.org/project/azure-multiapi-storage/ suggests `azure-multiapi-storage` might cover `azure-storage-blob`, but I don't know how to use it, and would rather not make an important dependency change at this late stage. I've raised work item https://dev.azure.com/msazuredev/AzureForOperators/_workitems/edit/992831 to track.

**Review by yanzhudd — APPROVED** [review]

File: src/aosm/azext_aosm/_configuration.py
Comment (bebound) [inline src/aosm/azext_aosm/_configuration.py:271]: It looks like this line is not compatible with 3.11.
> Change field default mutability check, allowing only defaults which are [hashable](https://docs.python.org/3/glossary.html#term-hashable) instead of any object which is not an instance of [dict](https://docs.python.org/3/library/stdtypes.html#dict), [list](https://docs.python.org/3/library/stdtypes.html#list) or [set](https://docs.python.org/3/library/stdtypes.html#set). (Contributed by Eric V. Smith in [bpo-44674](https://bugs.python.org/issue?@action=redirect&bpo=44674).)   --[What’s New In Python 3.11 — Python 3.12.0 documentation](https://docs.python.org/3/whatsnew/3.11.html#dataclasses)

This error is raised when run `az self-test --debug`:
```
  File "/opt/az/azcliextensions/aosm/azext_aosm/_configuration.py", line 267, in <module>
    @dataclass
     ^^^^^^^^^
  File "/opt/hostedtoolcache/Python/3.11.6/x64/lib/python3.11/dataclasses.py", line 1230, in dataclass
    return wrap(cls)
           ^^^^^^^^^
  File "/opt/hostedtoolcache/Python/3.11.6/x64/lib/python3.11/dataclasses.py", line 1220, in wrap
    return _process_class(cls, init, repr, eq, order, unsafe_hash,
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/hostedtoolcache/Python/3.11.6/x64/lib/python3.11/dataclasses.py", line 958, in _process_class
    cls_fields.append(_get_field(cls, name, type, kw_only))
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/hostedtoolcache/Python/3.11.6/x64/lib/python3.11/dataclasses.py", line 815, in _get_field
    raise ValueError(f'mutable default {type(f.default)} for field '
ValueError: mutable default <class 'azext_aosm._configuration.ArtifactConfig'> for field arm_template is not allowed: use default_factory
```

Comment (Cyclam) [reply src/aosm/azext_aosm/_configuration.py:271]: Are we expected to support Python 3.11? I can't run `azdev setup` against 3.11, and the pipelines for our PR didn't test at Python 3.11, as far as I can see.

I have a potential fix, but because of the above problems, testing and ensuring there aren't other occurrences of this problem (I expect there will be), is not possible.

**Pull request merged**
